### ISSUE TITLE
NativeCallback fastcall and thiscall broken on x86/Windows

### ISSUE THREAD

**Issue by oleavr** [frida/frida-gum#191]
Continuation of frida/frida#229. As explained by @mrgreywater:

Interceptor.replace in conjunction with NativeCallback frees too much stackspace after executing with the ABI 'fastcall' and 'thiscall'

Example msvc++ source to reproduce (Hit any key to execute all functions once): [FridaAbiTest.cpp](https://gist.github.com/mrgreywater/ad1d392dbd9b51a028e234f4a820e5fd)
Or as compiled binary (debug build -O0, no ASLR, with stack-checks): [download](http://www64.zippyshare.com/v/V32qd7E3/file.html)
Example Script for Frida: [FridaAbiTest.js](https://gist.github.com/mrgreywater/2a11eb29883785f1b5b1b7e3ae5f025f)

The example executable will run fine as long as Frida doesn't replace either the fastcall or thiscall function. As soon as that happens, the application crashes with a stack assertion (if those are activated in the compiler). If stack assertions are disabled, the application will usually just crash.

**Comment by 4got2think** [comment]
adding

```
#ifdef G_OS_WIN32
  switch (abi) {
	case FFI_THISCALL:
		  for (size_t argidx = 0; argidx < callback->cif.nargs; ++argidx)
		  {
			  if (argidx == 1)
				  break;
			  callback->cif.bytes -= callback->atypes[argidx]->size;
		  }
		  break;

	  case FFI_FASTCALL:
		  for (size_t argidx = 0; argidx < callback->cif.nargs; ++argidx)
		  {
			  if (argidx == 2)
				  break;
			  callback->cif.bytes -= callback->atypes[argidx]->size;
		  }
		  break;

	  default:
		  break;
  }
#endif
```


at https://github.com/frida/frida-gum/blob/master/bindings/gumjs/gumv8core.cpp#L2672 worked for me

hacky solution, you can think of something better for sure
it used to clean up the register arguments size too
don't know if it breaks any other code, it doesn't seem to

**Comment by pushad** [comment]
Is there a temporary solution to this bug? I'm running into the same issue.

**Comment by oleavr** [comment]
The temporary solution is to generate a callback yourself that calls the plain `NativeCallback`:
https://www.youtube.com/watch?v=9cr4gOPFN4o

The real solution is to solve this in [libffi](https://github.com/frida/libffi).

**Comment by 2567351318** [comment]
How to call a thiscall function

**Comment by oleavr** [comment]
This is now finally fixed in Frida 14.2.3, and libffi patches have been submitted upstream 🎉 Thanks a lot for reporting and helping narrow this one down!
